File: 0x0019_25.Reverse_Nodes_In_K-Group/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        // if is null, return null
        if (head==null) return head;
        if (k==1) return head;
        if (head.next==null) return head;
        // take at most k steps, find the end node
             
        int l=1;
        ListNode node=head;
        while (node.next!=null && l<k) {
            node=node.next;
            l+=1;
        }
        ListNode newHead=node;
        
        // if there is more recurse on the remaining nodes
        ListNode remain=null;
        if (l<k) return head;
        if (l>=k) {
            remain=node.next;
        }
        remain=reverseKGroup(remain,k);
           
        // reverse current segment 
        ListNode prev=head;
        ListNode cur=head.next;
       
        ListNode tail=head;
        tail.next=null;
        while (cur.next != null && cur!=newHead) {
            ListNode next=cur.next;
            cur.next=prev;
            prev=cur;
            cur=next;
        }
        cur.next=prev;
        
        // concat them and return head
        tail.next=remain;
        return newHead;    
    }
}

File: 0x0019_25.Reverse_Nodes_In_K-Group/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def reverseKGroup(self, head: "ListNode", k: "int") -> "ListNode":
        def getlen(node):
            i = 0
            while node:
                i += 1
                node = node.next
            return i

        def helper(node):
            if getlen(node) < k:
                return node, None, None
            p = None
            h = None
            t = None
            i = 0
            while i < k:
                n = node.next
                node.next = p
                if t is None:
                    t = node
                h = node
                p = node
                node = n
                i += 1
            return h, t, n

        nh = None
        nt = None
        while head:
            h, t, n = helper(head)
            if nh is None:
                nh = h
            if nt is not None:
                nt.next = h
            nt = t
            head = n

        return nh
